ipatests/test_webui/test_user.py

```diff
@@ -588,6 +864,164 @@ def test_life_cycles(self):
+
+        # add SSH key, move elsewhere and cancel.
+        self.add_sshkey_to_user(user.SSH_RSA, save=False)
+        self.assert_num_ssh_keys(1)
+        self.switch_to_facet('memberof_group')
+        self.dialog_button_click('cancel')
+        self.assert_num_ssh_keys(1)
+        self.undo_ssh_keys()
+
+        # add SSH key, move elsewhere and click reset button.
+        self.add_sshkey_to_user(user.SSH_RSA, save=False)
+        self.assert_num_ssh_keys(1)
+        self.switch_to_facet('memberof_group')
+        self.wait_for_request()
+        self.dialog_button_click('revert')
+        self.wait()
+        self.switch_to_facet('details')
+        self.assert_num_ssh_keys(0)
+
+        # add SSH key, move elsewhere and click save button.
+        self.add_sshkey_to_user(user.SSH_RSA, save=False)
+        self.assert_num_ssh_keys(1)
+        self.switch_to_facet('memberof_group')
+        self.wait()
+        self.dialog_button_click('save')
+        self.wait_for_request(n=4)
+        self.switch_to_facet('details')
+        self.assert_num_ssh_keys(1)
+        self.delete_user_sshkeys(navigate=False)
+
+        # add,save and delete RSA and DSA keys
+        keys = [user.SSH_RSA, user.SSH_DSA]
+
+        self.add_sshkey_to_user(keys)
+        self.assert_num_ssh_keys(2)
+        self.delete_user_sshkeys()
+        self.assert_num_ssh_keys(0)
+
+        # add RSA SSH keys with trailing space and "=" sign at the end
+        keys = [user.SSH_RSA+" ", user.SSH_RSA2+"="]
+
+        self.add_sshkey_to_user(keys)
+        self.assert_num_ssh_keys(2)
+        self.delete_user_sshkeys()
+        self.assert_num_ssh_keys(0)
+
+        # lets try to add empty SSH key (should fail)
+        self.add_sshkey_to_user('')
+        self.assert_last_error_dialog(EMPTY_MOD)
+        self.dialog_button_click('cancel')
+        self.undo_ssh_keys()
+
+        # try to add invalid SSH key
+        self.add_sshkey_to_user('invalid_key')
+        self.assert_last_error_dialog(INVALID_SSH_KEY)
+        self.dialog_button_click('cancel')
+        self.undo_ssh_keys()
+
+        # add duplicate SSH keys
+        self.add_sshkey_to_user(user.SSH_RSA)
+        self.add_sshkey_to_user(user.SSH_RSA, save=False)
+        self.facet_button_click('save')
+        self.assert_last_error_dialog(EMPTY_MOD)
+        self.dialog_button_click('cancel')
+
+        # test SSH key edit when user lacks write rights for related attribute
+        # see ticket 3800
+        self.add_record(user.ENTITY, [user.DATA2, user.DATA5])
+        self.add_sshkey_to_user(user.SSH_RSA, user.PKEY2)
+
+        self.logout()
+        self.init_app(user.PKEY5, '!!!@@@###$$$')
+
+        self.navigate_to_record(user.PKEY2, entity=user.ENTITY)
+
+        show_ssh_key_btn = self.find('div.widget .btn[name="ipasshpubkey-0"]',
+                                     By.CSS_SELECTOR)
+        show_ssh_key_btn.click()
+        ssh_key_dialog = self.find('.modal-dialog textarea', By.CSS_SELECTOR)
```

Comment:
The variable name is not precise. It is a textarea element and not a dialog. A bit of a nitpic: Maybe it could be:
ssh_key_element =  `self.find('textarea', By.CSS_SELECTOR, self.get_dialog())`

The `get_dialog` call should ensure that it uses consistent way to get the dialog and also it is the last dialog opened.